kv/kvserver: TestLargeUnsplittableRangeReplicate failed

kv/kvserver.TestLargeUnsplittableRangeReplicate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21115048?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21115048?buildTab=artifacts#/) on release-26.1 @ [95f5cc09c39ce567a622485f53b13f0822b95c80](https://github.com/cockroachdb/cockroach/commits/95f5cc09c39ce567a622485f53b13f0822b95c80):


```
=== RUN   TestLargeUnsplittableRangeReplicate
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/7fb080117a3da2d7cdf3576571d60489/logTestLargeUnsplittableRangeReplicate3175818455
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:668: DRPC is enabled (via override by TestingGlobalDRPCOption)
    replicate_queue_test.go:1880: 
        	Error Trace:	pkg/kv/kvserver/replicate_queue_test.go:1880
        	Error:      	Received unexpected error:
        	            	pq: none of the remaining voters [n2,s2] are legal additions to (n1,s1):1,(n5,s5):2,(n3,s3):4
        	Test:       	TestLargeUnsplittableRangeReplicate
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7fb080117a3da2d7cdf3576571d60489/logTestLargeUnsplittableRangeReplicate3175818455
--- FAIL: TestLargeUnsplittableRangeReplicate (3.88s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLargeUnsplittableRangeReplicate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-60472

--- a/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go
+++ b/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go
@@ -1231,17 +1231,24 @@ func rankedCandidateListForAllocation(
 	// count of those which are valid/not-full.
 	validCandidateStores := []roachpb.StoreDescriptor{}
 	for _, s := range candidateStores.Stores {
-		if !options.getDiskOptions().maxCapacityCheck(s) ||
-			!options.getIOOverloadOptions().allocateReplicaToCheck(
-				ctx,
-				s,
-				candidateStores,
-			) {
+		if !options.getDiskOptions().maxCapacityCheck(s) {
+			log.KvDistribution.VEventf(ctx, 3,
+				"not considering store s%d as a %s candidate: disk at or near capacity",
+				s.StoreID, targetType)
 			continue
 		}
-
+		if !options.getIOOverloadOptions().allocateReplicaToCheck(ctx, s, candidateStores) {
+			log.KvDistribution.VEventf(ctx, 3,
+				"not considering store s%d as a %s candidate: IO overloaded",
+				s.StoreID, targetType)
+			continue
+		}
 		if constraintsOK, _ := constraintsCheck(s); constraintsOK {
 			validCandidateStores = append(validCandidateStores, s)
+		} else {
+			log.KvDistribution.VEventf(ctx, 3,
+				"not considering store s%d as a %s candidate: failed constraints check",
+				s.StoreID, targetType)
 		}
 	}
 
@@ -1252,11 +1259,17 @@ func rankedCandidateListForAllocation(
 	for _, s := range validStoreList.Stores {
 		// Disregard all the stores that already have replicas.
 		if StoreHasReplica(s.StoreID, existingReplTargets) {
+			log.KvDistribution.VEventf(ctx, 3,
+				"not considering store s%d as a %s candidate: store already has a replica",
+				s.StoreID, targetType)
 			continue
 		}
 		// Unless the caller specifically allows us to allocate multiple replicas on
 		// the same node, we disregard nodes with existing replicas.
 		if !allowMultipleReplsPerNode && nodeHasReplica(s.Node.NodeID, existingReplTargets) {
+			log.KvDistribution.VEventf(ctx, 3,
+				"not considering store s%d as a %s candidate: node n%d already has a replica",
+				s.StoreID, targetType, s.Node.NodeID)
 			continue
 		}
 

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -35,6 +35,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness/livenesspb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvtestutils"
 	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
 	"github.com/cockroachdb/cockroach/pkg/raft/tracker"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -1757,11 +1758,12 @@ func TestLargeUnsplittableRangeReplicate(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.UnderStress(t, 38565)
-	skip.UnderRaceWithIssue(t, 38565)
+	skip.UnderDuressWithIssue(t, 38565)
 	skip.UnderShort(t, 38565)
-	skip.UnderDeadlockWithIssue(t, 38565)
+
 	ctx := context.Background()
+
+	testutils.SetVModule(t, "allocator_scorer=3")
 
 	const rangeMaxSize = 64 << 20
 	zcfg := zonepb.DefaultZoneConfig()
@@ -1799,8 +1801,14 @@ func TestLargeUnsplittableRangeReplicate(t *testing.T) {
 	_, err := db.Exec("create table t (i int primary key, s string)")
 	require.NoError(t, err)
 
-	_, err = db.Exec(`ALTER TABLE t EXPERIMENTAL_RELOCATE VALUES (ARRAY[1,2,3], 1)`)
-	require.NoError(t, err)
+	testutils.SucceedsSoon(t, func() error {
+		_, err := db.Exec(`ALTER TABLE t EXPERIMENTAL_RELOCATE VALUES (ARRAY[1,2,3], 1)`)
+		if kvtestutils.IsExpectedRelocateError(err) {
+			return err
+		}
+		require.NoError(t, err)
+		return nil
+	})
 	_, err = db.Exec(`ALTER TABLE t SPLIT AT VALUES (1)`)
 	require.NoError(t, err)
 	_, err = db.Exec(`ALTER TABLE t SPLIT AT VALUES (2)`)

